libraries/botbuilder/src/botFrameworkAdapter.ts

```diff
@@ -1151,36 +1149,37 @@ export class BotFrameworkAdapter extends BotAdapter implements IUserTokenProvide
      * @param res The response sent on error or connection termination.
      * @param logic The logic that will handle incoming requests.
      */
-    public async useWebSocket(req: WebRequest, res: WebResponse, logic: (context: TurnContext) => Promise<any>): Promise<void> {   
-        if (!logic) {
-            throw new Error('Streaming logic needs to be provided to `useWebSocket`');
-        }
-
+    public async useWebSocket(req: IncomingMessage, socket: Socket, head: Buffer, logic: (context: TurnContext) => Promise<any>): Promise<void> {   
         if (!this.webSocketFactory || !this.webSocketFactory.createWebSocket) {
             throw new Error('BotFrameworkAdapter must have a WebSocketFactory in order to support streaming.');
         }
 
-        this.logic = logic;
-
-        // Restify-specific check.
-        if (typeof((res as any).claimUpgrade) !== 'function') {
-            throw new Error('ClaimUpgrade is required for creating WebSocket connection.');
+        if (!logic) {
+            throw new Error('Streaming logic needs to be provided to `useWebSocket`');
         }
 
+        this.logic = logic;
+
         try {
             await this.authenticateConnection(req, this.settings.channelService);
         } catch (err) {
-            // Set the correct status code for the socket to send back to the channel.
-            res.status(StatusCodes.UNAUTHORIZED);
-            res.send(err.message);
+            // If the authenticateConnection call fails, send back the correct error code and close
+            // the connection.
+            if (typeof(err.message) === 'string' && err.message.toLowerCase().startsWith('unauthorized')) {
+                abortWebSocketUpgrade(socket, 401);
+            } else if (typeof(err.message) === 'string' && err.message.toLowerCase().startsWith(`'authheader'`)) {
+                abortWebSocketUpgrade(socket, 400);
+            } else {
+                abortWebSocketUpgrade(socket, 500);
+            }
```

Comment:
@carlosscastro I'm not happy with this code, but I'm not sure if there are better steps that can be taken here.

Node.js does not have HTTP Status Code-specific errors, so our underlying auth layers throw `new Error` with different message strings. Due to this, the code parses the message to attempt to determine the correct status code to send back before destroying the socket connection.

I have a test that covers the 401 Unauthorized route, and tested the other two routes manually.

Reply:
Ah, that hurts, but I understand. Can we improve the error reporting in the auth layers instead of here? Maybe we could create an abstraction that processes error messages and creates well defined errors + codes, and then the auth layer uses that in order to throw richer errors. Sounds like something that may not be possible without breaking backward compat though.

If it is not possible to move this logic to the auth layer, then I guess we are out of choices.

Reply:
Filed https://github.com/microsoft/botbuilder-js/issues/1445 for this. I've tentatively labeled it as a P2 for R7. Since this is internal, I don't want to block on this specific code.

I'd like to get this in so that I can have the package available for the WebChat DLS bug bash tomorrow.